src/adapter/src/coord/read_policy.rs

```diff
@@ -421,6 +425,39 @@ impl crate::coord::Coordinator {
         read_holds
     }
 
+    /// Attempt to acquire read holds on the indicated collections at the indicated `time`.
+    /// This is similar to [Self::acquire_read_holds], but if it is unable to acquire read holds at precisely
+    /// the specified time, then it errors out rather than acquiring read holds at a later time.
+    ///
+    /// The returned error contains those collection sinces that were later than the specified time.
+    pub(crate) fn acquire_precise_read_holds(
+        &mut self,
+        requested_time: Timestamp,
+        id_bundle: &CollectionIdBundle,
+    ) -> Result<ReadHolds<Timestamp>, Vec<(Antichain<Timestamp>, CollectionIdBundle)>> {
+        let read_holds = self.acquire_read_holds(requested_time, id_bundle);
+        let too_late = read_holds
+            .holds
+            .iter()
+            .filter_map(|(antichain, ids)| {
+                if antichain
+                    .iter()
+                    .all(|hold_time| *hold_time == requested_time)
+                {
+                    None
+                } else {
+                    Some((antichain.clone(), ids.clone()))
+                }
+            })
+            .collect_vec();
+        if !too_late.is_empty() {
+            self.release_read_holds(read_holds);
```

Comment:
Can we refactor these methods so that we don't actually send a message to compute/storage unless we're going to return `Ok(_)`.

Reply:
Done: b5b20de0f52a60a408cb560acd4f0fc64643609f
For this, I needed to add an argument to `acquire_read_holds` that indicates whether we want precise read holds, because that's where we can still decide not to actually acquire the read holds. This turned out nice, as it also allowed to collapse `acquire_read_holds` and `acquire_precise_read_holds` into one function, and the same for `acquire_read_holds_auto_cleanup` and `acquire_precise_read_holds_auto_cleanup`.